### PR TITLE
[c10d] Make reduce_scatter as a custom op

### PR DESCRIPTION
**Pull request by alanwaketan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79779
* #79777
* #79691
* #79688
* #79687
* #79686
* __->__ #79683
* #79669
* #79582

Summary:
This patch makes reduce_scatter as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_reduce_scatter_ops

### CODE REVIEW

File: torch/csrc/distributed/c10d/Ops.cpp [hunk 1]
```diff
@@ -121,5 +138,26 @@ c10::intrusive_ptr<ProcessGroup::Work> allgather(
...
+    const std::vector<std::vector<at::Tensor>>& input_tensors,
+    const ReduceScatterOptions& opts) {
+  static auto op = c10::Dispatcher::singleton()
+                       .findSchemaOrThrow("c10d::reduce_scatter_", "")
+                       .typed<c10::intrusive_ptr<::c10d::ProcessGroup::Work>(
+                           const std::vector<at::Tensor>&,
+                           const std::vector<std::vector<at::Tensor>>&,
+                           const c10::intrusive_ptr<::c10d::ProcessGroup>&,
+                           int64_t,
+                           int64_t)>();
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/Ops.cpp:153]: Would I be correct that the overhead of these two ops are negligible?

File: torch/csrc/distributed/c10d/init.cpp [hunk 2]
```diff
@@ -1178,23 +1178,29 @@ that adds a prefix to each key inserted to the store.
...
+              },
               py::arg("output_tensors"),
               py::arg("input_tensors"),
               py::arg("opts") = ::c10d::ReduceScatterOptions(),
               py::call_guard<py::gil_scoped_release>())
 
           .def(
               "reduce_scatter",
-              [](::c10d::ProcessGroup& pg,
+              [](const c10::intrusive_ptr<::c10d::ProcessGroup>& pg,
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/init.cpp:1195]: let's stay consistent on the naming? This one is using `pg`. The one above is using `self`.

**Review by mrshenli — APPROVED** [review]
LGTM! Same suggestions on tests as the PR below this one.

File: torch/csrc/distributed/c10d/Ops.cpp (hunk 1 above)
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/Ops.cpp:153]: It should be negligible. And then we cache it for the consecutive calls to minimize the overhead too.

File: torch/csrc/distributed/c10d/init.cpp (hunk 2 above)
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/init.cpp:1195]: I believe this pg is user defined, and the above is really `self`. That's why there is this inconsistency.

**Comment by alanwaketan** [comment]
Thanks Shen for approving this pull request. The CI failure is unrelated.

File: torch/csrc/distributed/c10d/init.cpp (hunk 2 above)
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/init.cpp:1195]: Let me follow up on #80246.

**Pull request closed without merging**